### ISSUE TITLE
Improve ConcurrentBag<T> GetEnumerator performance

### ISSUE THREAD

**Issue by MarkusSintonen** [dotnet/corefx#2335]
Currently `ConcurrentBag<T>` `GetEnumerator()` causes list copying everytime enumerator is accessed eg. through `foreach`. This causes strain on the GC and list copying is also heavy operation on large bags.

There should be alternative methods to enumarate the bag contents or implement `GetEnumerator()` so that it does not cause list copying. I have presented here 3 possbile solutions to this problem.

**Solution 1)**
Add `ForEach` method to `ConcurrentBag<T>`:

``` CSharp
public void ForEach(Func<T, bool> action)
{
    // Short path if the bag is empty
    if (_headList == null)
        return;

    bool lockTaken = false;
    try
    {
        FreezeBag(ref lockTaken);

        ThreadLocalList currentList = _headList;
        while (currentList != null) // Go through thread lists
        {
            Node currentNode = currentList._head;
            while (currentNode != null) // Go through thead specific linked list
            {
                if (!action(currentNode._value))
                    return; // User wanted to break the operation
                currentNode = currentNode._next;
            }
            currentList = currentList._nextList;
        }
    }
    finally
    {
        UnfreezeBag(lockTaken);
    }
}
```

`ForEach` function takes user specified `Func<T,bool>` delegate. This delegate takes the bag item as an argument and returns true/false depending on whether the user wants to continue or stop the enumeration operation.

Here `ForEach` operation would lock the bag so that other threads would need to wait for the operation to finish. Therefore results of this operation would represent the bag as a moment-in-time snapshot.

**Solution 2)** 
Add `GetEnumerableSlim` function:

``` CSharp
public IEnumerable<T> GetEnumerableSlim()
{
    ThreadLocalList currentList = _headList;
    while (currentList != null)
    {
        Node currentNode = currentList._head;
        while (currentNode != null)
        {
            yield return currentNode._value;
            currentNode = currentNode._next;
        }
        currentList = currentList._nextList;
    }
}
```

This would require an additional change to the internal class `Node`:

``` CSharp
internal class Node
{
    public Node(T value)
    {
        _value = value;
    }
    public readonly T _value;
    public volatile Node _next; // This needs to be marked as volatile
    public Node _prev;
}
```

`GetEnumerableSlim` would enumerate the bag contents concurrently without locking the bag. The proposed implementation is thread safe from memory model point of view. The fields `ConcurrentBag<T>._headList`, `ThreadLocalList._head` and `Node._next` are all volatile variables. Therefore all the threads always have the newest value available. In addition the `Node._value` reference/value is immutable. `GetEnumerableSlim` is safe to use concurrently.

Of course here the enumerator would miss some of the concurrent additions/removals from the bag. Also enumerator could contain modification after the `GetEnumerator` was called. Similiar behaviour is also used in `ConcurrentDictionary<K,T>`, `GetEnumerator` which may also miss some of the concurrent changes or contain new changes after the `GetEnumerator` was called. Therefore the `GetEnumerableSlim` is consistent with other concurrent collections and their enumerator implementations. (It would be intresting to know why `ConcurrentBag<T>` `GetEnumerator` was originally implementeted so that it copies the bag contents to a list where as the other concurrent collections enumerate the contents concurrently eg. `ConcurrentQueue<T>`.)

**Solution 3)**
Replace `GetEnumerator` with `GetEnumeratorSlim` implementation. This solution would totally eliminate the risk of performance degradation due to list allocations. But this might not be possible if the existing code relies on the fact that `GetEnumerator` returns moment-in-time representation of the bag.

**Comment by stephentoub** [comment]
Thanks for opening the issue, @MarkusSintonen.

As long as we could demonstrate it doesn't negatively impact performance of existing code/usage (e.g. with the additional volatile you mention), I personally think (2) would be a worthwhile addition.  One of the common usage patterns of ConcurrentBag is to do lots of manipulations of it in parallel, e.g. building up a set of results, and then enumerating it in a non-parallel context; today that requires taking a bunch of locks and copying the contents into a ```List<T>```, and an API like this would address that case, in addition to allowing for concurrent usage (though I'm not sure how common it is to want to enumerate a bag while it's being manipulated concurrently).  From a naming perspective, I'm not a big fan of "GetEnumeratorSlim()", but I'm sure we could come up with something reasonable.

I'd be concerned with (1), as we'd be invoking arbitrary code while holding a bunch of locks and could easily lead to deadlock.  And (3) is likely a non-starter due to the behavioral change you mention, which I expect could have a fairly significant impact.

Curious to hear other opinions about (2) in particular.

@terrajobst, FYI.

**Comment by MarkusSintonen** [comment]
Thank you for the comment @stephentoub, I don't know how common it is to enumerate contents of the bag concurrently but I once encountered situation where this would have been useful. In this case bunch of threads wrote to the, lets call it also bag, while single thread checked constantly the contents for specific condition. Here list allocations with locking would have degraged the performance.

**Comment by MarkusSintonen** [comment]
I tested whether the volatile on `Node._next` has any add/remove performance implications. I found out that it had no implications. I tested the implementation by running 8 threads each adding and removing 12500 items to/from bag. Test was repeated 10000 times and average of timings was taken. Timings were measured on how long it takes to add and remove 12500 items per thread.

I think this is expected since `Node._next` is accessed only by a single thread when `GetEnumerableSlim` is not called. Therefore there is no performance implications to the existing code.

**Comment by stephentoub** [comment]
> I tested whether the volatile on Node._next has any add/remove performance implications.

Thanks.  We'd probably want to double check that, too, on some other architectures, like ARM.

cc: @ericeil

**Comment by MarkusSintonen** [comment]
Yes, I tested it on Intel i7-3630QM if that matters anyway.

**Comment by karelz** [comment]
Seems that it has been addressed in the linked PR. Feel free to reopen if you disagree.

**Comment by stephentoub** [comment]
> Seems that it has been addressed in the linked PR. Feel free to reopen if you disagree.

No, it hasn't been addressed.  That PR was just a corner case, avoiding an allocation when the bag is empty.  This issue is mainly focused on enabling entirely allocation free enumeration, which I think is a good thing to do, e.g. adding a method like:
```C#
public Enumerator DangerousGetEnumerator();
```
that would a) return a struct (to avoid the ```IEnumerator<T>``` allocation, and more importantly b) iterate the bag in place rather than copying it to an array first.  These changes would make it possible to iterate through the contents of the bag, but doing so would only be safe when there was no concurrent usage of the bag, e.g. in a situation where concurrent activity builds up the contents of the bag, and then once all that activity is done, its contents can be enumerated (hence the "Dangerous" prefix, but we could come up with something else if desired).

The aforementioned method would enable using MoveNext/Current explicitly on the enumerator.  If we wanted to enable foreach/LINQ/etc., we could instead have a struct-based enumerable that had a GetEnumerator which returned the same struct-based enumerator, e.g.
```C#
public Enumerable DangerousGetEnumerable();
```
or something like that.
